R/check.R

```diff
@@ -1,90 +1,186 @@
   if (is.null(value)) {
     if (required) {
-      rlang::abort(paste0("argument ", name, " is not a epirange"),
+      rlang::abort(paste0("argument ", name, " is not a datelike"),
         name = name, value = value, class = "invalid_argument"
       )
     }
+  } else if (is_date_like(value)) {
     return()
-  }
-  if (is_epirange_like(value)) {
-    return()
-  }
-  if (!is.list || !all(sapply(sapply(value, is_epirange_like)))) {
-    rlang::abort(paste0("argument ", name, " is not a epirange"),
+  } else if (is.list(value) && length(value) > 0) {
+    for (i in seq_along(value)) {
+      check_single_date_param(paste0(name, "[", i, "]"), value[[i]], required = TRUE)
+    }
+  } else {
+    rlang::abort(paste0("argument ", name, " is not a datelike"),
       name = name, value = value, class = "invalid_argument"
     )
   }
 }
 
-check_single_string_param <- function(name, value, required = TRUE) {
-  if ((required &&
-    !(!is.null(value) &&
-      is.character(value) &&
-      length(value) == 1)) ||
-    (!required &&
-      !(is.null(value) ||
-        (
-          is.character(value) && length(value) == 1
-        )))) {
-    rlang::abort(paste0("argument ", name, " is not a single string"),
+#' Return TRUE if value is a character or numeric or EpiRange
+#' TODO: add Date support
+is_epirange_like <- function(value) {
+  is.character(value) ||
+    is.numeric(value) ||
+    inherits(value, "EpiRange")
+}
+
+#' Check if param is a character or numeric or EpiRange
+check_single_epirange_param <- function(name, value, required = TRUE) {
+  if (is.null(value)) {
+    if (required) {
+      rlang::abort(paste0("argument ", name, " is not an epirange_like"),
+        name = name, value = value, class = "invalid_argument"
+      )
+    }
+  } else if (is_epirange_like(value)) {
+    return()
+  } else {
+    rlang::abort(paste0("argument ", name, " is not an epirange_like"),
       name = name, value = value, class = "invalid_argument"
     )
   }
 }
 
-check_single_int_param <- function(name, value, required = TRUE) {
-  if ((required &&
-    !(!is.null(value) &&
-      is.numeric(value) &&
-      length(value) == 1)) ||
-    (!required &&
-      !(is.null(value) || (is.numeric(value) && length(value) == 1)))) {
-    rlang::abort(paste0("argument ", name, " is not a single integer"),
+#' Check if param is a character or numeric or EpiRange
+check_epirange_param <- function(name, value, required = TRUE) {
+  if (is.null(value)) {
+    if (required) {
+      rlang::abort(paste0("argument ", name, " is not an epirange_like"),
+        name = name, value = value, class = "invalid_argument"
+      )
+    }
+    return()
+  } else if (is_epirange_like(value)) {
+    return()
+  } else if (all(vapply(value, is.list, logical(1L))) && length(value) > 0) {
+    for (i in seq_along(value)) {
+      check_single_epirange_param(paste0(name, "[", i, "]"), value[[i]], required = TRUE)
+    }
+  } else if (sum(names(value) %in% c("from", "to")) >= 2) {
```

Comment:
polish: eliminate the from&to part here since it was removed from the single epirange-like above.

Reply:
This is my round-about way of detecting when someone gives us a vector of epiranges, which becomes a mangled list:
```
r$> c(epirange(1, 2), epirange(3, 4))
$from
[1] 1

$to
[1] 2

$from
[1] 3

$to
[1] 4
```
Is there a better way to handle that?